### PR TITLE
Migrate SSSS to use symmetric algorithm

### PR DESCRIPTION
**Pull request by uhoreg**
includes the changes from https://github.com/matrix-org/matrix-js-sdk/pull/1228

### CODE REVIEW

**Comment by uhoreg** [comment]
fixes https://github.com/vector-im/riot-web/issues/12508

This should be basically ready for review, but needs some figuring out about how it interfaces with higher levels, which probably needs some react-sdk implementation.

File: src/crypto/index.js [hunk 1]
```diff
@@ -413,19 +439,42 @@ Crypto.prototype.bootstrapSecretStorage = async function({
...
+                // if we already have a backup key, use the same key as the
+                // secret storage key
                 logger.log("Secret storage default key not found, using key backup key");
-                const opts = {
-                    pubkey: keyBackupInfo.auth_data.public_key,
-                };
+                const opts = {};
+
+                // FIXME: ask for recovery passphrase/key
+                const backupKey = "foobar";
```
Comment (uhoreg) [inline src/crypto/index.js]: I don't think we have a callback yet for asking for the backup key, so one may need to be added.

File: src/crypto/index.js [hunk 2]
```diff
@@ -388,10 +388,36 @@ Crypto.prototype.bootstrapSecretStorage = async function({
...
+                ssssKeys[keyId] = keyData;
+            }
+            return key;
+        };
+
     try {
-        const inStorage = !setupNewSecretStorage &&
-            await this._crossSigningInfo.isStoredInSecretStorage(this._secretStorage);
+        const decryptionKeys =
+              await this._crossSigningInfo.isStoredInSecretStorage(this._secretStorage);
```
Comment (foldleft) [inline src/crypto/index.js]: I worry that waiting on a promise might return control to the main loop, leaving the callbacks above in place

Comment (foldleft) [reply src/crypto/index.js]: not that I'm sure there's a better way to do this

File: src/crypto/index.js [hunk 3]
```diff
@@ -413,19 +439,42 @@ Crypto.prototype.bootstrapSecretStorage = async function({
                     { authUploadDeviceSigningKeys },
                 );
             }
+        } else if (!(Object.values(decryptionKeys).some(
+            info => info.algorithm === SECRET_STORAGE_ALGORITHM_V1_AES,
+        ))) {
+            // we already have cross-signing keys, but they're encrypted using
+            // the old algorithm
+            logger.log("Switching to symmetric");
```
Comment (foldleft) [inline src/crypto/index.js]: this probably merits a `logger.warn`, plus more context

File: src/crypto/index.js (hunk 1 above)
Comment (foldleft) [reply src/crypto/index.js]: file a task if you're leaving this unfinished

Comment (uhoreg) [reply src/crypto/index.js]: Yeah, this is part of the reason why this PR is marked as draft.  It'll probably need some figuring out with the react side how to do it properly.

**Comment by turt2live** [comment]
Taking riot-web review off this as it's my understanding it's not completely ready yet.

**Pull request merged**